### ISSUE TITLE
Linux: `empty_string` triggers in cases where the left hand side isn't a string.

### ISSUE THREAD

**Issue by CraigSiemens** [realm/SwiftLint#4249]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

On `main` (fd7afedfcfe73443a95041fdc51c31d071b97910), the `empty_string` rule seems to check if the `rhs` of an equality operator is an empty string literal. This can give false positives when the `lhs` isn't a string.

In our case it's occurring when running in a github workflow on linux. It fails in a test using Nimble matchers comparing to an empty string. In this case the suggestion doesn't make sense since `.isEmpty` isn't available on the type on the `lhs`.

```
expect(outputText) == ""
```

Looking at the implementation, it may also give false positives for types that are `StringLiteralConvertible`. The suggestions of using `isEmpty` also wouldn't be possible in that case.

Though, maybe the lint warning message should be changed from assuming the lhs is a string.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
path/to/file: warning:
Empty String Violation: Prefer checking `isEmpty` over comparing `string` to an empty string literal. (empty_string)
Done linting! Found 0 violations, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.49.1 (main)
* Installation method used (Homebrew, CocoaPods, building from source, etc)? building from source
* Paste your configuration file:

```yml
opt_in_rules:
  - empty_string
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcodebuild -version`)? n/a
* Do you have a sample that shows the issue?

I'd suspect this reproduces it but I don't have an environment where I can run it myself.
```
echo "expect(outputText) == \"\"" | swiftlint lint --no-cache --use-stdin --enable-all-rules
```

**Comment by SimplyDanny** [comment]
The rule indeed behaves a bit different now, but has never considered the type of the expression on the left-hand side. For example, `a == ""` would have triggered previously even though `a` was not a string. It did not trigger, however, if the left-hand side was a function call. The new implementation does not care for a call or a simple identifier or something else.

That said, the new implementation is not worse than the previous. In your example, it is causing a false-positive. On the other hand, it now correctly reports `f() == ""` in case the return type of `f()` supports `.isEmpty`.
